cdap-app-templates/cdap-program-report/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala

```diff
@@ -143,20 +151,76 @@ object ReportGenerationHelper {
     reportIdDir.append(Constants.LocationName.SUCCESS_FILE).createNew()
   }
 
+  /**
+    * Generates a summary of the report with the information from the report generation request and the DataFrame
+    * containing the report details, then writes the summary to the given location.
+    *
+    * @param request the report generation request
+    * @param df the DataFrame containing report details
+    * @param reportIdDir the location to write the summary to
+    */
+  private def writeSummary(request: ReportGenerationRequest, df: DataFrame, reportIdDir: Location): Unit = {
+    val namespaces = mutable.HashSet[NamespaceAggregate]()
+    // group the report details by namespace, and then collect the count and the corresponding unique namespaces
+    df.groupBy(Constants.NAMESPACE).count.collect.foreach(r => namespaces +=
+      new NamespaceAggregate(r.getAs[String](Constants.NAMESPACE), r.getAs[Long](COUNT_COL)))
+    // group the report details by artifact information including artifact name, version and scope,
+    // and then collect the count and the corresponding unique artifact information
+    val artifacts =  mutable.HashSet[ArtifactAggregate]()
+    df.groupBy(Constants.ARTIFACT_NAME, Constants.ARTIFACT_VERSION, Constants.ARTIFACT_SCOPE).count.collect
+      .foreach(r => artifacts += new ArtifactAggregate(r.getAs[String](Constants.ARTIFACT_NAME),
+        r.getAs[String](Constants.ARTIFACT_VERSION), r.getAs[String](Constants.ARTIFACT_SCOPE),
+        r.getAs[Long](COUNT_COL)))
+    // aggregate the report details into a row with the min, max, and average of duration,
+    // and with the min and max of start
+    val aggRow = df.agg(min(df(Constants.DURATION)).as("minDuration"), max(df(Constants.DURATION)).as("maxDuration"),
+      avg(df(Constants.DURATION)).as("avgDuration"), min(df(Constants.START)).as("minStart"),
+      max(df(Constants.START)).as("maxStart")).first
```

Comment:
first because there is only one row? is this the only way to create a row ? cant you directly create row ?